lib/src/common/disposable.dart

```diff
@@ -21,25 +21,45 @@ import 'package:meta/meta.dart';
 import 'package:w_common/src/common/disposable_manager.dart';
 import 'package:w_common/src/common/managed_stream_subscription.dart';
 
-// ignore: one_member_abstracts
-abstract class _Disposable {
-  Future<Null> dispose();
-}
-
-class InternalDisposable implements _Disposable {
+class InternalDisposable implements SimpleDisposable {
   Disposer _disposer;
+  final Completer<Null> _didDispose = new Completer<Null>();
+  bool _isDisposing = false;
 
   InternalDisposable(this._disposer);
 
   @override
   Future<Null> dispose() {
-    var disposeFuture = _disposer != null ? _disposer() : null;
-    _disposer = null;
-    if (disposeFuture == null) {
-      return new Future.value();
+    if (isDisposed) {
+      return null;
```

Comment:
That is in an async block so the null return value will be returned as the value in a `Future` instance. What if this dispose method is used in a sync block:

```dart
internalDisposable.dispose().then((_) => print('I will throw because the dispose method returned null'));
```